src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs

```diff
@@ -43,253 +54,202 @@ protected AbstractUseAutoPropertyAnalyzer()
         protected abstract SyntaxNode GetFieldNode(TFieldDeclaration fieldDeclaration, TVariableDeclarator variableDeclarator);
 
         protected abstract void RegisterIneligibleFieldsAction(
-            List<AnalysisResult> analysisResults, HashSet<IFieldSymbol> ineligibleFields,
-            Compilation compilation, CancellationToken cancellationToken);
-
-        protected sealed override void InitializeWorker(AnalysisContext context)
-            => context.RegisterSemanticModelAction(AnalyzeSemanticModel);
+            ConcurrentSet<IFieldSymbol> ineligibleFields, SemanticModel semanticModel, SyntaxNode codeBlock, CancellationToken cancellationToken);
 
-        private void AnalyzeSemanticModel(SemanticModelAnalysisContext context)
-        {
-            var cancellationToken = context.CancellationToken;
-            var semanticModel = context.SemanticModel;
+        protected abstract void RegisterNonConstructorFieldWrites(
+            ConcurrentDictionary<IFieldSymbol, ConcurrentSet<SyntaxNode>> fieldWrites, SemanticModel semanticModel, SyntaxNode codeBlock, CancellationToken cancellationToken);
 
-            // Don't even bother doing the analysis if the user doesn't even want auto-props.
-            var option = context.GetAnalyzerOptions().PreferAutoProperties;
-            if (!option.Value)
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSymbolStartAction(context =>
             {
-                return;
-            }
-
-            var analysisResults = new List<AnalysisResult>();
-            var ineligibleFields = new HashSet<IFieldSymbol>();
+                var namedType = (INamedTypeSymbol)context.Symbol;
+                if (namedType.TypeKind is not TypeKind.Class and not TypeKind.Struct and not TypeKind.Module)
+                    return;
 
-            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
-            AnalyzeCompilationUnit(context, root, analysisResults);
+                var analysisResults = new ConcurrentQueue<AnalysisResult>();
+                context.RegisterSyntaxNodeAction(context => AnalyzePropertyDeclaration(context, namedType, analysisResults), PropertyDeclarationKind);
```

Comment:
question.  will this only analyze PropertyDecls *directly* within the containing symbol, or anything deep within a nested type?

Reply:
Only the directly contained property decls for the outermost type. If you had a nested type, then callbacks for property decls within the nested type will be made to the syntax node action registered within SymbolStart callback for the nested type symbol.

Reply:
that's cool.  can you show me where teh code for that lives @mavasani i want to deeply understand the invariants there.